### PR TITLE
XAG resubstitution for multiplicative complexity

### PR DESCRIPTION
**Pull request by eletesta**

### CODE REVIEW

File: include/mockturtle/algorithms/resubstitution.hpp [hunk 1]
```diff
@@ -177,13 +182,13 @@ class node_mffc_inside
...
-        auto const& p = ntk.get_node( f );
+    ntk.foreach_fanin( n, [&]( const auto& f ) {
+      auto const& p = ntk.get_node( f );
 
-        ntk.decr_fanout_size( p );
-        if ( ntk.fanout_size( p ) == 0 )
-          counter += node_deref_rec( p );
-      });
+      ntk.decr_fanout_size( p );
+      if ( ntk.fanout_size( p ) == 0 )
```
Comment (msoeken) [inline include/mockturtle/algorithms/resubstitution.hpp]: Please use explicit scope `{...}` for single statements.

File: include/mockturtle/algorithms/resubstitution.hpp [hunk 2]
```diff
@@ -195,14 +200,14 @@ class node_mffc_inside
...
+      auto const& p = ntk.get_node( f );
 
-        auto v = ntk.fanout_size( p );
-        ntk.incr_fanout_size( p );
-        if ( v == 0 )
-          counter += node_ref_rec( p );
-      });
+      auto v = ntk.fanout_size( p );
+      ntk.incr_fanout_size( p );
+      if ( v == 0 )
```
Comment (msoeken) [inline include/mockturtle/algorithms/resubstitution.hpp]: See above.

File: include/mockturtle/algorithms/resubstitution.hpp [hunk 3]
```diff
@@ -361,25 +362,22 @@ class default_resub_functor
...
 
-  std::optional<signal> operator()( node const& root, uint32_t required, uint32_t max_inserts, uint32_t num_mffc, uint32_t& last_gain ) const
+  std::optional<signal> operator()( node const& root, TT care, uint32_t required, uint32_t max_inserts, uint32_t num_mffc, uint32_t& last_gain ) const
   {
     /* The default resubstitution functor does not insert any gates
        and consequently does not use the argument `max_inserts`. Other
        functors, however, make use of this argument. */
     (void)max_inserts;
-
+    assert(kitty::is_const0(~care));
```
Comment (msoeken) [inline include/mockturtle/algorithms/resubstitution.hpp]: Inconsistent formatting (spaces).

File: include/mockturtle/algorithms/resubstitution.hpp [hunk 4]
```diff
@@ -616,10 +634,16 @@ class resubstitution_impl
...
 
     /* simulate the collected divisors */
-    call_with_stopwatch( st.time_simulation, [&]() { simulate( leaves ); });
+    call_with_stopwatch( st.time_simulation, [&]() { simulate( leaves ); } );
+
+    auto care = kitty::create<TT>( leaves.size() );
+    if ( ps.use_dont_cares )
+      care = ~satisfiability_dont_cares( ntk, leaves, 12u );
+    else 
+      care = (~care);
```
Comment (msoeken) [inline include/mockturtle/algorithms/resubstitution.hpp]: `care = ~care`

**Comment by hriener** [comment]
The changes related to updating fanout may affect the quality and runtime of _all_ resubstitution algorithm. It may make sense to rerun our resubstitution experiments before merging these changes.

**Comment by hriener** [comment]
Thanks. Seems to work now. We may want to merge this PR before PR #244. Is the PR ready?

**Comment by eletesta** [comment]
I think we can merge the PR —QoR is similar to DATE 2020

On 7 Feb 2020, at 20:14, Heinz Riener <notifications@github.com<mailto:notifications@github.com>> wrote:


Thanks. Seems to work now. We may want to merge this PR before PR #244<https://github.com/lsils/mockturtle/pull/244>. Is the PR ready?

**Pull request merged**